File: src/engine/deliberation.ts
/**
 * 译前推敲（与 Python tav2/deliberation.py 对齐）：
 * 多方位评估候选术语，查证为证据，高置信自动采纳，其余进审批队列。
 * 查证通过 EvidenceProvider 注入；引擎默认无查证（dsh 侧接 tool-web）。
 */

import type { EngineConfig } from './config'
import type { ProjectDB } from './db'
import type { Generate } from './llm'
import { extractJson, extractJsonArray } from './llm'
import { deliberationBatchPrompt, deliberationEvalPrompt } from './prompts'
import { activateEntries, type MemoryEntry } from './memory'

export interface SearchResult {
  title: string
  url: string
  snippet: string
}

/** 查证来源：输入查询词，返回最多 3 条证据。可注入 dsh tool-web 或离线假实现。 */
export type EvidenceProvider = (query: string) => Promise<SearchResult[]>

/**
 * 普通词典词判定（方案 B）：全小写拉丁词且不在专名白名单 → 普通词，不推敲不锁定。
 * 含大写字母（专名）、或命中 scan.extra_proper_nouns（游戏特有设定词）的不算。
 */
export function isCommonDictionaryWord(source: string, extraProperNouns: string[]): boolean {
  const s = String(source ?? '').trim()
  if (!/[A-Za-z]/.test(s)) return false
  if (s !== s.toLowerCase()) return false
  const proper = new Set(
    (extraProperNouns ?? []).map((w) => String(w).trim().toLowerCase()).filter(Boolean),
  )
  return !proper.has(s.toLowerCase())
}

export interface DeliberationStats {
  evaluated: number
  auto_locked: number
  pending_approval: number
  failed: number
}

export function evidenceText(results: SearchResult[]): string {
  if (!results || results.length === 0) return '（无查证结果）'
  return results.slice(0, 3)
    .map((r) => `- ${r.title || ''}（${r.url || ''}）：${r.snippet || ''}`)
    .join('\n')
}

function hasLatin(text: string): boolean {
  return /[A-Za-z]/.test(text)
}

function queryFor(source: string): string {
  return hasLatin(source) ? `"${source}" 翻译 中文 译名` : `${source} 译名`
}

function contextText(cand: Record<string, unknown>): string {
  const samples = Array.isArray(cand.samples) ? cand.samples.map(String) : []
  if (samples.length > 0) return samples.slice(0, 3).join('\n')
  return '（无语境样本）'
}

/** 用确认过的世界书条目按文本激活，拼成推敲背景（常驻 + 命中）。 */
function worldbookContext(worldbook: MemoryEntry[], text: string): string {
  const [constants, hits] = activateEntries(worldbook, text)
  const items = [...constants, ...hits]
  if (items.length === 0) return ''
  return items.map((e) => `【${String(e.title ?? '')}】${String(e.content ?? '')}`).join('\n')
}

interface Decision {
  target: string
  confidence: string
  rationale: string
  collision: string
}

function decisionFrom(data: Record<string, unknown>): Decision {
  return {
    target: String(data.target ?? '').trim(),
    confidence: String(data.confidence ?? 'low'),
    rationale: String(data.rationale ?? ''),
    collision: String(data.collision ?? ''),
  }
}

function applyDecision(
  db: ProjectDB,
  cand: Record<string, unknown>,
  data: Record<string, unknown>,
  stats: DeliberationStats,
): void {
  const source = String(cand.source ?? '').trim()
  const decision = decisionFrom(data)
  if (!decision.target) {
    stats.failed += 1
    return
  }
  // S13：落库前按 source 清理旧候选（含占位行 (source,'')），upsert 冲突键是
  // (source,target)，不清理则改译名会新增行、候选反复叠加翻倍。
  db.clearCandidateTermsBySource(source)
  db.upsertTerm(
    source,
    decision.target,
    String(cand.category ?? ''),
    'candidate',
    decision.confidence,
    decision.rationale,
  )
  stats.evaluated += 1
  const row = db.termBySourceTarget(source, decision.target)
  if (decision.confidence === 'high' && !decision.collision) {
    const id = Number(row?.id)
    if (Number.isFinite(id) && db.decideTerm(id, 'locked')) stats.auto_locked += 1
  } else {
    db.addApproval('term', {
      source,
      target: decision.target,
      confidence: decision.confidence,
      rationale: decision.rationale,
    })
    stats.pending_approval += 1
  }
}

async function evaluateOne(
  generate: Generate,
  db: ProjectDB,
  cand: Record<string, unknown>,
  evidence: EvidenceProvider,
  stats: DeliberationStats,
  worldbook: MemoryEntry[],
): Promise<void> {
  const source = String(cand.source ?? '').trim()
  try {
    const results = await evidence(queryFor(source))
    const wb = worldbookContext(worldbook, `${source}\n${contextText(cand)}`)
    const result = await generate.generate({
      system: deliberationEvalPrompt(),
      messages: [{
        role: 'user',
        content: [
          `候选：${source}`,
          `类别：${String(cand.category ?? '')}`,
          `出现次数：${String(cand.frequency ?? '?')}`,
          `出现语境：\n${contextText(cand)}`,
          `世界书背景：${wb ? '\n' + wb : '（无）'}`,
          `考量维度：用典、玩梗、文化、韵律、双关、短习俚等，综合判定译名。`,
          `查证证据：\n${evidenceText(results)}`,
        ].join('\n'),
      }],
    })
    applyDecision(db, cand, extractJson(result.text), stats)
  } catch {
    stats.failed += 1
  }
}

async function buildCandidateBlock(
  index: number,
  cand: Record<string, unknown>,
  evidence: EvidenceProvider,
  worldbook: MemoryEntry[],
): Promise<string> {
  const source = String(cand.source ?? '').trim()
  const results = await evidence(queryFor(source))
  const wb = worldbookContext(worldbook, `${source}\n${contextText(cand)}`)
  return [
    `${index}. 候选：${source}`,
    `类别：${String(cand.category ?? '')}`,
    `出现次数：${String(cand.frequency ?? '?')}`,
    `出现语境：\n${contextText(cand)}`,
    `世界书背景：${wb ? '\n' + wb : '（无）'}`,
    `考量维度：用典、玩梗、文化、韵律、双关、短习俚等，综合判定译名。`,
    `查证证据：\n${evidenceText(results)}`,
  ].join('\n')
}

async function evaluateBatch(
  generate: Generate,
  db: ProjectDB,
  batch: Array<Record<string, unknown>>,
  evidence: EvidenceProvider,
  stats: DeliberationStats,
  worldbook: MemoryEntry[],
): Promise<void> {
  const rows: string[] = []
  for (let i = 0; i < batch.length; i++) {
    rows.push(await buildCandidateBlock(i + 1, batch[i]!, evidence, worldbook))
  }
  try {
    const result = await generate.generate({
      system: deliberationBatchPrompt(rows.join('\n\n')),
      messages: [{ role: 'user', content: rows.join('\n\n') }],
    })
    const decisions = extractJsonArray(result.text)
    const bySource = new Map<string, Record<string, unknown>>()
    for (const d of decisions) {
      bySource.set(String(d.source ?? '').trim(), d)
    }
    for (const cand of batch) {
      const decision = bySource.get(String(cand.source ?? '').trim())
      if (!decision) {
        stats.failed += 1
        continue
      }
      applyDecision(db, cand, decision, stats)
    }
  } catch {
    // 批量失败逐条兜底，保证不丢候选（与 Python 一致）
    for (const cand of batch) await evaluateOne(generate, db, cand, evidence, stats, worldbook)
  }
}

/** 评估候选：查证 + LLM 多方位决策，结果写回 DB。返回统计。 */
export async function evaluateCandidates(
  generate: Generate,
  db: ProjectDB,
  cfg: EngineConfig,
  candidates?: Array<Record<string, unknown>>,
  evidence: EvidenceProvider = async () => [],
): Promise<DeliberationStats> {
  const source = candidates ?? db.pendingTerms()
  const items = source
    .map((c) => ({ ...c }))
    .filter((c) => String(c.source ?? '').trim())
    // 方案 B：普通词典词一律不推敲、不锁定（含 NSFW 高频词），白名单与专名照常。
    .filter((c) => !isCommonDictionaryWord(String(c.source ?? ''), cfg.scan.extraProperNouns))
  const stats: DeliberationStats = { evaluated: 0, auto_locked: 0, pending_approval: 0, failed: 0 }
  const batchSize = Math.max(1, Math.floor(cfg.deliberation.batchSize))
  // 推敲背景：确认过的世界书条目（常驻 + 按候选语境命中）。
  const worldbook = db.loadWorldbook()
  for (let start = 0; start < items.length; start += batchSize) {
    const batch = items.slice(start, start + batchSize)
    if (batch.length <= 1) {
      for (const cand of batch) await evaluateOne(generate, db, cand, evidence, stats, worldbook)
    } else {
      await evaluateBatch(generate, db, batch, evidence, stats, worldbook)
    }
  }
  return stats
}

File: python/tav2/deliberation.py
"""译前推敲：多方位评估候选术语，联网查证为证据，高置信自动采纳。"""

from __future__ import annotations

from typing import Any

from tav2.llm import BaseLLM, LLMError
from tav2.prompts import DELIBERATION_BATCH_PROMPT, DELIBERATION_EVAL_PROMPT, render
from tav2.websearch import search


def _evidence_text(results: list[dict[str, str]]) -> str:
    if not results:
        return "（无查证结果）"
    lines = []
    for r in results[:3]:
        lines.append(f"- {r.get('title') or ''}（{r.get('url') or ''}）：{r.get('snippet') or ''}")
    return "\n".join(lines)


def evaluate_candidates(
    llm: BaseLLM,
    db: Any,
    cfg: dict[str, Any],
    candidates: list[dict[str, Any]] | None = None,
) -> dict[str, int]:
    """评估候选：联网查证 + LLM 多方位决策；结果写回 DB。

    返回统计 {evaluated, auto_locked, pending_approval, failed}。
    """

    if candidates is None:
        rows = db.conn.execute(
            "SELECT * FROM terms WHERE status='candidate' ORDER BY id"
        ).fetchall()
        candidates = [dict(r) for r in rows]
    stats = {"evaluated": 0, "auto_locked": 0, "pending_approval": 0, "failed": 0}

    items = [dict(c) for c in candidates if str(c.get("source") or "").strip()]
    batch_size = max(1, int(cfg.get("deliberation", {}).get("batch_size", 10)))
    # 推敲背景：确认过的世界书条目（常驻 + 按候选语境命中）。
    worldbook = db.load_worldbook()
    for start in range(0, len(items), batch_size):
        batch = items[start : start + batch_size]
        if len(batch) <= 1:
            for cand in batch:
                _evaluate_one(llm, db, cfg, cand, stats, worldbook)
        else:
            _evaluate_batch(llm, db, cfg, batch, stats, worldbook)
    return stats


def _evaluate_one(
    llm: BaseLLM,
    db: Any,
    cfg: dict[str, Any],
    cand: dict[str, Any],
    stats: dict[str, int],
    worldbook: list[dict[str, Any]],
) -> None:
    """逐条评估：联网查证 + 单条决策。"""

    source = str(cand.get("source") or "").strip()
    context = _context_text(cand)
    query = f'"{source}" 翻译 中文 译名' if re_search_latin(source) else f"{source} 译名"
    results = search(cfg, db, query)
    prompt = render(DELIBERATION_EVAL_PROMPT.template)
    wb = _worldbook_context(worldbook, f"{source}\n{context}")
    user = (
        f"候选：{source}\n"
        f"类别：{cand.get('category') or ''}\n"
        f"出现次数：{cand.get('frequency') or '?'}\n"
        f"出现语境：\n{context}\n"
        f"世界书背景：{wb if wb else '（无）'}\n"
        f"考量维度：用典、玩梗、文化、韵律、双关、短习俚等，综合判定译名。\n"
        f"查证证据：\n{_evidence_text(results)}"
    )
    try:
        data = llm.chat_json(prompt, user)
    except (LLMError, ValueError, TypeError):
        stats["failed"] += 1
        return
    _apply_decision(db, cand, data, stats)


def _evaluate_batch(
    llm: BaseLLM,
    db: Any,
    cfg: dict[str, Any],
    batch: list[dict[str, Any]],
    stats: dict[str, int],
    worldbook: list[dict[str, Any]],
) -> None:
    """批量评估：一次请求评估整批候选；失败时逐条兜底，保证不丢候选。"""

    rows: list[str] = []
    for idx, cand in enumerate(batch, start=1):
        source = str(cand.get("source") or "").strip()
        query = f'"{source}" 翻译 中文 译名' if re_search_latin(source) else f"{source} 译名"
        results = search(cfg, db, query)
        wb = _worldbook_context(worldbook, f"{source}\n{_context_text(cand)}")
        rows.append(
            f"{idx}. 候选：{source}\n"
            f"类别：{cand.get('category') or ''}\n"
            f"出现次数：{cand.get('frequency') or '?'}\n"
            f"出现语境：\n{_context_text(cand)}\n"
            f"世界书背景：{wb if wb else '（无）'}\n"
            f"考量维度：用典、玩梗、文化、韵律、双关、短习俚等，综合判定译名。\n"
            f"查证证据：\n{_evidence_text(results)}"
        )
    prompt = render(DELIBERATION_BATCH_PROMPT.template)
    user = "\n\n".join(rows)
    try:
        decisions = llm.chat_json_array(prompt, user)
    except (LLMError, ValueError, TypeError):
        for cand in batch:
            _evaluate_one(llm, db, cfg, cand, stats, worldbook)
        return
    by_source = {
        str(d.get("source") or "").strip(): d
        for d in decisions
        if isinstance(d, dict)
    }
    for cand in batch:
        decision = by_source.get(str(cand.get("source") or "").strip())
        if decision is None:
            stats["failed"] += 1
            continue
        _apply_decision(db, cand, decision, stats)


def _apply_decision(
    db: Any,
    cand: dict[str, Any],
    data: dict[str, Any],
    stats: dict[str, int],
) -> None:
    """把单条决策落库：候选更新 + 高置信自动锁定或进审批队列。"""

    source = str(cand.get("source") or "").strip()
    target = str(data.get("target") or "").strip()
    confidence = str(data.get("confidence") or "low")
    rationale = str(data.get("rationale") or "")
    collision = str(data.get("collision") or "")
    if not target:
        stats["failed"] += 1
        return
    db.upsert_term(
        source,
        target,
        str(cand.get("category") or ""),
        status="candidate",
        confidence=confidence,
        evidence=rationale,
    )
    stats["evaluated"] += 1
    term_id = db.conn.execute(
        "SELECT id FROM terms WHERE source=? AND target=?", (source, target)
    ).fetchone()
    if confidence == "high" and not collision:
        if term_id is not None:
            db.decide_term(int(term_id["id"]), "locked")
        stats["auto_locked"] += 1
    else:
        db.add_approval(
            "term",
            {
                "source": source,
                "target": target,
                "confidence": confidence,
                "rationale": rationale,
            },
        )
        stats["pending_approval"] += 1


def _context_text(cand: dict[str, Any]) -> str:
    samples = cand.get("samples") or []
    if samples:
        return "\n".join(str(s) for s in samples[:3])
    return "（无语境样本）"


def _worldbook_context(worldbook: list[dict[str, Any]], text: str) -> str:
    """用确认过的世界书条目按文本激活，拼成推敲背景（常驻 + 命中）。"""

    import re

    constants: list[dict[str, Any]] = []
    hits: list[dict[str, Any]] = []
    for e in worldbook:
        if e.get("kind") == "constant":
            constants.append(e)
            continue
        for kw in e.get("keywords") or []:
            kw = str(kw).strip()
            if not kw:
                continue
            if re.search(r"[A-Za-z]", kw):
                pattern = r"(?<![A-Za-z0-9_])" + re.escape(kw) + r"(?![A-Za-z0-9_])"
                if re.search(pattern, text, re.IGNORECASE):
                    hits.append(e)
                    break
            elif kw in text:
                hits.append(e)
                break
    items = constants + hits
    if not items:
        return ""
    return "\n".join(f"【{e.get('title') or ''}】{e.get('content') or ''}" for e in items)


def re_search_latin(text: str) -> bool:
    import re

    return bool(re.search(r"[A-Za-z]", text))
